[DataGrid] Performance: avoid layout invalidation

romgrk opened the pull request:
This PR is based on some of the work in #12023, split here to make review/merge easier.

Changes:
 - Use `position: absolute` rows to avoid layout invalidation
 - Remove `virtualScrollerRenderZone`, not needed anymore since sticky
 - Compute the left offset on the fly rather than storing in the state
 - Consider no colspan info as meaning `{ colSpan: 1 }`, to avoid expensive computations on every row render

cherniavskii reviewed (commented):
Thanks for extracting it to a separate PR!

Overall, I'm not sure the absolute row positioning is worth it (see the dynamic row height comment below).
Other changes make sense to me 👍🏻

romgrk commented on packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:
```diff
@@ -258,97 +238,85 @@ export const useGridVirtualScroller = () => {
+    const rowProps = rootProps.slotProps?.row;
+    const rowModels = params.rows ?? currentPage.rows;
 
-    const { focusedCellColumnIndexNotInRange, renderedColumns } = getRenderedColumns(
```
The focused cell logic is way more complex than it needs to be, haven't had time to refactor until now.

romgrk commented on packages/x-data-grid/src/hooks/features/columns/useGridColumns.tsx:
```diff
@@ -96,6 +96,7 @@ export function useGridColumns(
 
       apiRef.current.setState(mergeColumnsState(columnsState));
       apiRef.current.publishEvent('columnsChange', columnsState.orderedFields);
+      apiRef.current.updateRenderContext?.();
```
It's defined in `useGridVirtualScroller` which isn't a top-level hook, it's a hook used in `GridVirtualScroller`. So it doesn't get initialized first. Not sure why `useGridVirtualScroller` was implemented that way. And yes this is called before the function is defined sometimes, but in those cases it doesn't matter that it's not called because it's during the 1st render, and the render context is updated right after the 1st render.

This should probably be refactored to move the hook where it should be, but I haven't had a reason to do the refactor other than it would be cleaner.

romgrk commented:
I split the original PR incorrectly, to prove that `position: absolute` is a good change I also need the rest of the work in #12023. I've split the offsets, colspan & cell focus refactor into yet another PR: #12280.

Ignore this one for now.
